qiskit/circuit/instruction.py

```diff
@@ -440,15 +440,28 @@ def copy(self, name=None):
         """
         cpy = self.__deepcopy__()
 
-        if name:
+        if name is not None:
             cpy.name = name
         return cpy
 
     def __deepcopy__(self, memo=None):
+        if memo is None:
+            memo = {}
+
+        obj_id = id(self)
+        if obj_id in memo:
+            return memo[obj_id]
+
         cpy = copy.copy(self)
-        cpy._params = copy.copy(self._params)
-        if self._definition:
+        memo[obj_id] = cpy
+
+        cpy._params = copy.deepcopy(self._params, memo)
+
+        if getattr(self, "_definition", None) is not None:
```

Comment:
This seems to have extra unnecessary steps. What was the problem with the current check?
```suggestion
        if self._definition:
```